helmfile.d/charts/gatekeeper/templates/policies/restrict-pod-disruption-budgets.rego

```diff
@@ -0,0 +1,181 @@
+    [pdb.metadata.name, pdb.spec.maxUnavailable],
+    )
+}
+
+# Reject PDB if minAvailable does not allow at least 1 pod disruption
+violation[{"msg": msg}] {
+    input.review.object.kind == "PodDisruptionBudget"
+    pdb := input.review.object
+
+    pdb.spec.minAvailable
+
+    pod_controller_group_kind := pod_controller_groups_kinds[_]
+    objs := [controllers | controllers := data.inventory.namespace[pdb.metadata.namespace][pod_controller_group_kind.group][pod_controller_group_kind.kind]]
+    obj := objs[_][_]
+
+    not mismatched_selector(pdb, obj)
+
+    not_valid_pdb_min_available(obj, pdb)
+    not replica_set_under_deployment(obj)
+
+    msg := sprintf(
+    "PodDisruptionBudget rejected: %v <%v> has %v replica(s) but PodDisruptionBudget <%v> has minAvailable of %v, minAvailable should always be lower than replica(s), and not used when replica(s) is set to 1.",
+    [obj.kind, obj.metadata.name, obj.spec.replicas, pdb.metadata.name, pdb.spec.minAvailable],
+    )
+}
+
+# Reject pod controller if connected PDBs maxUnavailable does not allow at least 1 pod disruption
+violation[{"msg": msg}] {
+    input.review.object.kind == pod_controller_groups_kinds[_].kind
+    obj := input.review.object
+    not replica_set_under_deployment(obj)
+
+    pdb := data.inventory.namespace[obj.metadata.namespace]["policy/v1"].PodDisruptionBudget[_]
+
+    pdb.spec.maxUnavailable
+
+    not mismatched_selector(pdb, obj)
+
+    not_valid_pdb_max_unavailable(pdb)
+    msg := sprintf(
+    "%v rejected: %v <%v> has been selected by PodDisruptionBudget <%v> but has maxUnavailable of %v, only positive integers or percentages are allowed for maxUnavailable",
+    [obj.kind, obj.kind, obj.metadata.name, pdb.metadata.name, pdb.spec.maxUnavailable],
+    )
+}
+
+# Reject pod controller if connected PDBs minAvailable does not allow at least 1 pod disruption
+violation[{"msg": msg}] {
+    input.review.object.kind == pod_controller_groups_kinds[_].kind
+    obj := input.review.object
+    not replica_set_under_deployment(obj)
+
+    pdb := data.inventory.namespace[obj.metadata.namespace]["policy/v1"].PodDisruptionBudget[_]
+
+    pdb.spec.minAvailable
+
+    not mismatched_selector(pdb, obj)
+
+    not_valid_pdb_min_available(obj, pdb)
+    msg := sprintf(
+    "%v rejected: %v <%v> has %v replica(s) but PodDisruptionBudget <%v> has minAvailable of %v, minAvailable should always be lower than replica(s), and not used when replica(s) is set to 1.",
+    [obj.kind, obj.kind, obj.metadata.name, obj.spec.replicas, pdb.metadata.name, pdb.spec.minAvailable],
+    )
+}
+
+# The type of pod controller to validate
+pod_controller_groups_kinds := [
+    {"group": "apps/v1", "kind": "Deployment"},
+    {"group": "apps/v1", "kind": "StatefulSet"},
+    {"group": "apps/v1", "kind": "ReplicaSet"},
+    {"group": "v1", "kind": "ReplicationController"}
+]
```

Comment:
You only use the group here in 1 place (below), might want to split these into separate functions, one for group and one for kind.

https://github.com/elastisys/compliantkubernetes-apps/blob/8d6c09e4021c9233498844e31b1d633cbd1e4e13/helmfile.d/charts/gatekeeper/templates/policies/restrict-pod-disruption-budgets.rego#L26

Or do they need to be together to make sure it's the same object? If I understand it correctly it should take one object at a time so even if the information is fetched from two functions that shouldn't be a problem.

Reply:
The idea was to ensure that it just uses these pairs of group and kind. This was the easiest way that I could think of to do this, but there are probably other ways.
I could also just have two separate lists of groups and kinds and then let it go through all combinations. I assume that is slightly less efficient, but probably not significantly.

IMO the current version feels nice to read and understand which groups and kinds belong to each other. But I think we should use the version that most people find easiest to read and understand. So I'm ok with changing it if that is what you and others want.

Reply:
Okay, I would be fine with leaving it but I don't think this reads super well when referencing this `pod_controller_group_kind.group`

Reply:
I will keep it like this for now then.